File: clojure/ch02/ex2.3.clj
(ns sicp.ch02 (:use clojure.contrib.test-is))
(use '[clojure.contrib.except :only (throw-if)])

; Old sqrt calculation from ex. 1.46:

(defn square [x] (* x x))

(defn average [x y]
  (/ (+ x y) 2))

(def tolerance 0.0001)

(defn good-enough? [guess x]
  (< (Math/abs (- (square guess) x)) tolerance))

(defn improve [guess x]
  (average guess (/ x guess)))
  
(defn iterative-improve [good-enough? improve]
  (fn [guess] 
    (loop [guess2 guess]
      (if (good-enough? guess2)
        guess2
        (recur (improve guess2))))))

(defn sqrt [x]
  ((iterative-improve
    #(good-enough? % x) #(improve % x)) 1.0))

; Points, segments, etc. from ex. 2.2

(defn make-point [x y] [x y])

(defn x-point [p] (get p 0))
(defn y-point [p] (get p 1))
  
(defn equal-points? [p1 p2]
  (and (= (x-point p1) (x-point p2)) (= (y-point p1) (y-point p2))))

(defn make-segment [start end] [start end])
  
(defn start-segment [s] (get s 0))
(defn end-segment [s]   (get s 1))

(defn midpoint-segment [s]
  (let [x1 (x-point (start-segment s))
        y1 (y-point (start-segment s))
        x2 (x-point (end-segment s))
        y2 (y-point (end-segment s))]
      (make-point (/ (+ x1 x2) 2) (/ (+ y1 y2) 2))))

(defn print-point [p] (println (format "(%f,%f)" (x-point p) (y-point p))))

; Generic line length calculator:

(defn length-line [p1 p2]
  (sqrt (+ (square (- (x-point p2) (x-point p1))) 
           (square (- (y-point p2) (y-point p1))))))
  
; Rectangles: an implementation that uses 4 points.

; counterclockwise - but doesn't assume horizontal-vertical
; p4 -- p3
; |     |
; p1 -- p2 
(defn make-rect [p1 p2 p3 p4] [p1 p2 p3 p4])

(defn height-rect [r]
  (let [p4 (get r 3)
        p1 (get r 0)]
        (length-line p4 p1)))

(defn width-rect [r]
  (let [p2 (get r 1)
        p1 (get r 0)]
        (length-line p2 p1)))
  
; pass in the functions to calculate the height and width, so we can configure
; them using different rectangle implementations.

(defn area-rect [height-calc width-calc r]
  (* (height-calc r) (width-calc r)))
  
(defn perimeter-rect [height-calc width-calc r]
  (* 2 (+ (height-calc r) (width-calc r))))

(def zero-zero (make-point  0.0 0.0))
(def zero-two  (make-point  0.0 2.0))
(def two-zero  (make-point  2.0 0.0))
(def two-two   (make-point  2.0 2.0))
(def m-one-one (make-point -1.0 1.0))
(def one-one   (make-point  1.0 1.0))

(defn close-enough? [x y]
  (< (Math/abs (- x y)) tolerance))

(def sqrt-two (sqrt 2.0))

(deftest test-rect-with-4-points
  (def rect1 (make-rect zero-zero two-zero two-two zero-two))
  (def rect2 (make-rect zero-zero one-one zero-two m-one-one))
  (is (= (close-enough? (width-rect     rect1) 2) true))
  (is (= (close-enough? (height-rect    rect1) 2) true))
  (is (= (close-enough? (perimeter-rect height-rect width-rect rect1) 8) true))
  (is (= (close-enough? (area-rect      height-rect width-rect rect1) 4) true))
  (is (= (close-enough? (width-rect     rect2) sqrt-two) true))
  (is (= (close-enough? (height-rect    rect2) sqrt-two) true))
  (is (= (close-enough? (perimeter-rect height-rect width-rect rect2) (* 4 sqrt-two)) true))
  (is (= (close-enough? (area-rect      height-rect width-rect rect2) 2) true)))

; Rectangles: 2nd implementation that uses 2 segments. 1st is assumed to be the
; more vertical; its length is taken to be the height. The 2nd is assumed to be
; the more horizontal; its length is the width.
; Area and perimeter methods unchanged.

; start point for both segments must be the same.
(defn make-rect-segments [s1 s2]
  (throw-if (= false (equal-points? (start-segment s1) (start-segment s2))) 
            (str "must use segments with the same starting points." (start-segment s1) (start-segment s2)))
  [s1 s2])

(defn height-rect-segment [r]
  (let [p2 (start-segment (get r 0))
        p1 (end-segment   (get r 0))]
        (length-line p1 p2)))

(defn width-rect-segment [r]
  (let [p2 (start-segment (get r 1))
        p1 (end-segment   (get r 1))]
        (length-line p1 p2)))
  
(deftest test-rect-with-2-segments
  (def rect3 (make-rect-segments 
    (make-segment zero-zero two-zero) (make-segment zero-zero zero-two)))
  (def rect4 (make-rect-segments 
    (make-segment zero-zero one-one) (make-segment zero-zero m-one-one)))
  (is (= (close-enough? (width-rect-segment  rect3) 2) true))
  (is (= (close-enough? (height-rect-segment rect3) 2) true))
  (is (= (close-enough? (perimeter-rect height-rect-segment width-rect-segment rect3) 8) true))
  (is (= (close-enough? (area-rect      height-rect-segment width-rect-segment rect3) 4) true))

  (is (= (close-enough? (width-rect-segment  rect4) sqrt-two) true))
  (is (= (close-enough? (height-rect-segment rect4) sqrt-two) true))
  (is (= (close-enough? (perimeter-rect height-rect-segment width-rect-segment rect4) (* 4 sqrt-two)) true))
  (is (= (close-enough? (area-rect      height-rect-segment width-rect-segment rect4) 2) true)))

(run-tests)

File: scala/ch02/ex2.3.scala
// Points and segments from ex. 2.2

type point = Tuple2[Double,Double]

def makePoint (x: Double, y: Double) = (x, y)

def xPoint (p: point) = p._1
def yPoint (p: point) = p._2
  
type segment = Tuple2[point,point]

def makeSegment(start: point, end: point) = (start, end)
  
def startSegment (s: segment) = s._1
def endSegment   (s: segment) = s._2

def midpointSegment (s: segment) = {
  val x1 = xPoint (startSegment(s))
  val y1 = yPoint (startSegment(s))
  val x2 = xPoint (endSegment(s))
  val y2 = yPoint (endSegment(s))
  makePoint ((x1 + x2) / 2, (y1 + y2) / 2)
}

def printPoint (p: point) = format ("(%f,%f)\n", xPoint(p), yPoint(p))

def equalPoints (p1: point, p2: point) = 
    xPoint(p1) == xPoint(p2) && yPoint(p1) == yPoint(p2)

// Sqrt from ex 1.46

val tolerance = 0.00001

def square(x: Double) = x*x

def iterativeImprove(goodEnough: (Double) => Boolean, improve: (Double) => Double) = {
  def iterate(guess2: Double): Double = goodEnough(guess2) match {
    case true => guess2
    case _ => iterate(improve(guess2))
  }
  (guess:Double) => iterate(guess)
}

def average(x: Double, y: Double) = (x + y) / 2.0

def goodEnough(guess: Double, x: Double) = math.abs(square(guess) - x) < tolerance

def improve(guess: Double, x: Double) = average(guess, x / guess)
  
def sqrt(x: Double) = iterativeImprove(
  (guess:Double) => goodEnough(guess, x), 
  (guess:Double) => improve(guess, x))(1.0)
    
// Test rectangle implementations.

val zeroZero = makePoint (0.0, 0.0)
val zeroTwo  = makePoint (0.0, 2.0)
val twoZero  = makePoint (2.0, 0.0)
val oneOne   = makePoint (1.0, 1.0)
val twoTwo   = makePoint (2.0, 2.0)
val mOneOne  = makePoint (-1.0, 1.0)

import org.scalatest._
import org.scalatest.matchers._

// Generic line length calculator:

def lengthLine (p1: point, p2: point) = sqrt(
    square (xPoint(p2) - xPoint(p1)) + square (yPoint(p2) - yPoint(p1)))

// Rectangles: an implementation that uses 4 points.

type rect = Tuple4[point,point,point,point]

// counterclockwise - but doesn't assume horizontal-vertical
// p4 -- p3
// |     |
// p1 -- p2 
def makeRect (p1: point, p2: point, p3: point, p4: point) = (p1, p2, p3, p4)

def heightRect (r: rect) = lengthLine (r._4, r._1)

def widthRect (r: rect) = lengthLine (r._2, r._1)

// pass in the functions to calculate the height and width, so we can configure
// them using different rectangle implementations.

def areaRect[R] (heightCalc: (R) => Double, widthCalc: (R) => Double, r: R) =
  heightCalc(r) * widthCalc(r)

def perimeterRect[R] (heightCalc: (R) => Double, widthCalc: (R) => Double, r: R) =
  2 * (heightCalc(r) + widthCalc(r))

val sqrtTwo = sqrt(2.0)

import org.scalatest._
import org.scalatest.matchers._

object rectWith4PointsSpec extends Spec with ShouldMatchers {
  describe ("Rectangle represented by 4 points") {
    it ("should calculate the area and perimeter") {
      val rect1 = makeRect (zeroZero, twoZero, twoTwo,  zeroTwo)
      widthRect     (rect1) should be (2.0 plusOrMinus tolerance)
      heightRect    (rect1) should be (2.0 plusOrMinus tolerance)
      perimeterRect (heightRect, widthRect, rect1) should be (8.0 plusOrMinus tolerance)
      areaRect      (heightRect, widthRect, rect1) should be (4.0 plusOrMinus tolerance)

      val rect2 = makeRect (zeroZero, oneOne,  zeroTwo, mOneOne)
      widthRect     (rect2) should be (sqrtTwo plusOrMinus tolerance)
      heightRect    (rect2) should be (sqrtTwo plusOrMinus tolerance)
      perimeterRect (heightRect, widthRect, rect2) should be ((4.0*sqrtTwo) plusOrMinus tolerance)
      areaRect      (heightRect, widthRect, rect2) should be (2.0 plusOrMinus tolerance)
    }
  }
}
rectWith4PointsSpec execute

// Rectangles: 2nd implementation that uses 2 segments. 1st is assumed to be the
// more vertical; its length is taken to be the height. The 2nd is assumed to be
// the more horizontal; its length is the width.
// Area and perimeter methods unchanged.

type rectSegment = Tuple2[segment,segment]

// start point for both segments must be the same.
def makeRectSegments (s1: segment, s2: segment) = {
  if (equalPoints (startSegment(s1), startSegment(s2)) == false) 
    throw new RuntimeException ("must use segments with the same starting points.")
  else
    (s1, s2)
}

def heightRectSegment (r: rectSegment) = {
  val p2 = startSegment (r._1)
  val p1 = endSegment   (r._1)
  lengthLine (p1, p2)
}

def widthRectSegment (r: rectSegment) = {
  val p2 = startSegment (r._2)
  val p1 = endSegment   (r._2)
  lengthLine (p1, p2)
}

object rectWith2SegmentsSpec extends Spec with ShouldMatchers {
  describe ("Rectangle represented by 2 segments") {
    it ("should calculate the area and perimeter") {
      val rect3 = makeRectSegments (makeSegment(zeroZero, twoZero), makeSegment(zeroZero, zeroTwo))
      widthRectSegment  (rect3) should be (2.0 plusOrMinus tolerance)
      heightRectSegment (rect3) should be (2.0 plusOrMinus tolerance)
      perimeterRect     (heightRectSegment, widthRectSegment, rect3) should be (8.0 plusOrMinus tolerance)
      areaRect          (heightRectSegment, widthRectSegment, rect3) should be (4.0 plusOrMinus tolerance)

      val rect4 = makeRectSegments (makeSegment(zeroZero, oneOne), makeSegment(zeroZero, mOneOne))
      widthRectSegment  (rect4) should be (sqrtTwo plusOrMinus tolerance)
      heightRectSegment (rect4) should be (sqrtTwo plusOrMinus tolerance)
      perimeterRect     (heightRectSegment, widthRectSegment, rect4) should be ((4.0*sqrtTwo) plusOrMinus tolerance)
      areaRect          (heightRectSegment, widthRectSegment, rect4) should be (2.0 plusOrMinus tolerance)
    }
  }
}
rectWith2SegmentsSpec execute
